core-rust/state-manager/src/limits.rs

```diff
@@ -0,0 +1,179 @@
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+use node_common::config::limits::VertexLimitsConfig;
+use radix_engine::transaction::ExecutionMetrics;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum VertexLimitsExceeded {
+    TransactionsCount,
+    TransactionsSize,
+    ExecutionCostUnitsConsumed,
+    SubstateReadSize,
+    SubstateReadCount,
+    SubstateWriteSize,
+    SubstateWriteCount,
+}
+
+pub struct VertexLimitsTracker {
+    pub total_transactions_count: usize,
+    pub total_transactions_size: usize,
+    pub total_execution_cost_units_consumed: usize,
+    pub total_substate_read_size: usize,
+    pub total_substate_read_count: usize,
+    pub total_substate_write_size: usize,
+    pub total_substate_write_count: usize,
+}
+
+impl VertexLimitsTracker {
+    pub fn new() -> Self {
+        Self {
+            total_transactions_count: 0,
+            total_transactions_size: 0,
+            total_execution_cost_units_consumed: 0,
+            total_substate_read_size: 0,
+            total_substate_read_count: 0,
+            total_substate_write_size: 0,
+            total_substate_write_count: 0,
+        }
+    }
+
+    pub fn check_pre_execution(
+        &self,
+        limits_config: &VertexLimitsConfig,
+        transaction_size: usize,
+    ) -> Result<(), VertexLimitsExceeded> {
+        if self.total_transactions_count + 1 > limits_config.max_total_transactions_count {
+            return Err(VertexLimitsExceeded::TransactionsCount);
+        }
+
+        if self.total_substate_read_size + transaction_size
+            > limits_config.max_total_transactions_size
+        {
+            return Err(VertexLimitsExceeded::TransactionsSize);
+        }
+
+        Ok(())
+    }
+
+    pub fn check_post_execution(
+        &self,
+        limits_config: &VertexLimitsConfig,
+        execution_metrics: &ExecutionMetrics,
+    ) -> Result<(), VertexLimitsExceeded> {
+        if self.total_execution_cost_units_consumed
```

Comment:
I assume we're safe from overflows here due to limits in the engine?

Reply:
Right now, yes.

Reply:
I've refactor to `remaining_`, so this is not an issue anymore.